flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+			if (random.nextBoolean()) {
+				long namespace = namespaceEntry.getKey();
+				removedNamespaces.add(namespace);
+				for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+					int key = keyEntry.getKey();
+					if (random.nextBoolean()) {
+						stateMap.remove(key, namespace);
+					} else {
+						assertEquals(keyEntry.getValue(), stateMap.removeAndGetOld(key, namespace));
+					}
+					totalSize--;
+					assertEquals(totalSize, stateMap.size());
+					assertEquals(totalSize, stateMap.totalSize());
+				}
+			}
+		}
+
+		for (long namespace : removedNamespaces) {
+			referenceStates.remove(namespace);
+			assertEquals(0, stateMap.sizeOfNamespace(namespace));
+			assertFalse(stateMap.getKeys(namespace).iterator().hasNext());
+		}
+
+		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
+		verifyState(referenceStates, stateMap);
+
+		stateMap.close();
+		assertEquals(0, stateMap.size());
+		assertEquals(0, stateMap.totalSize());
+		// all spaces should be free
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+		assertTrue(stateMap.isClosed());
+	}
+
+	/**
+	 *  Tests copy-on-write contracts.
+	 */
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testCopyOnWriteContracts() throws IOException {
+		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
+		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+		// do not remove states physically when get, put, remove and snapshot
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(
+			keySerializer,
+			namespaceSerializer,
+			stateSerializer,
+			spaceAllocator,
+			0,
+			1.0f);
+
+		StateSnapshotTransformer<String> transformer = new StateSnapshotTransformer<String>() {
+			@Nullable
+			@Override
+			public String filterOrTransform(@Nullable String value) {
+				if (value == null) {
+					return null;
+				}
+				int op = value.hashCode() % 3;
+				switch (op) {
+					case 0:
+						return null;
+					case 1:
+						return value + "-transform";
+					default:
+						return value;
+				}
+			}
+		};
+
+		// map to store expected states, namespace -> key -> state
+		Map<Long, Map<Integer, String>> referenceStates = new HashMap<>();
+		int totalStateSize = 0;
+		int totalSizeIncludingLogicalRemovedKey = 0;
+		int totalLogicallyRemovedKey = 0;
+		int totalSpaceNumber = 0;
+
+		// take snapshot 1 which is an empty snapshot
+		Map<Long, Map<Integer, String>> expectedSnapshot1 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot1 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+		assertEquals(1, stateMap.getHighestRequiredSnapshotVersion());
+		assertEquals(1, stateMap.getSnapshotVersions().size());
+		assertEquals(true, stateMap.getSnapshotVersions().contains(1));
```

Comment:
could be simplified to `assertTrue`

Reply:
Or by using Hamcrest matchers: `assertThat(stateMap.getSnapshotVersions(), contains(1))`